### PR TITLE
Add buffer limit option for SignalR Stateful Reconnect

### PR DESCRIPTION
**Pull request by BrennanConroy**
Allows the user to configure buffer limits on server and client.

### CODE REVIEW

File: src/SignalR/common/SignalR.Common/src/Protocol/SequenceMessage.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+/// <summary>
+/// Represents the restart of the sequence of messages being sent. <see cref="SequenceId"/> is the starting ID of messages being sent, which might be duplicate messages.
+/// </summary>
+public sealed class SequenceMessage : HubMessage
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="SequenceMessage"/> class.
+    /// </summary>
+    /// <param name="sequenceId">Specifies the ID messages being received from now on should start at.</param>
```
Comment (adityamandaleeka) [inline src/SignalR/common/SignalR.Common/src/Protocol/SequenceMessage.cs:14]: Wording is a bit awkward. How about something like

```suggestion
    /// <param name="sequenceId">Specifies the starting ID for messages that will be received from this point onward.</param>
```

File: src/SignalR/clients/csharp/Client.Core/src/HubConnectionOptions.cs [hunk 2]
```diff
@@ -23,4 +23,10 @@ internal sealed class HubConnectionOptions
     /// Configures KeepAliveInterval for the <see cref="HubConnection" />.
     /// </summary>
     public TimeSpan? KeepAliveInterval { get; set; }
+
+    public const int DefaultMessageBufferSize = 100_000;
```
Comment (adityamandaleeka) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnectionOptions.cs:27]: Can you include the units here (at least as a comment)?

File: src/SignalR/server/Core/src/HubConnectionContextOptions.cs [hunk 3]
```diff
@@ -34,4 +34,9 @@ public class HubConnectionContextOptions
     /// Gets or sets the maximum parallel hub method invocations.
     /// </summary>
     public int MaximumParallelInvocations { get; set; } = 1;
+
+    /// <summary>
+    /// Gets or sets the maximum bytes to buffer per connection when using stateful reconnect.
+    /// </summary>
+    public long MessageBufferSize { get; set; } = 100_000;
```
Comment (adityamandaleeka) [inline src/SignalR/server/Core/src/HubConnectionContextOptions.cs:41]: Personally I like having units in names, but if this is the name that was decided during API review, I won't push back.

**Review by adityamandaleeka — APPROVED** [review]

File: src/SignalR/server/Core/src/HubConnectionContextOptions.cs (hunk 3 above)
Comment (adityamandaleeka) [reply src/SignalR/server/Core/src/HubConnectionContextOptions.cs:41]: Oh wait, that's tomorrow 😆

File: src/SignalR/common/Shared/MessageBuffer.cs [hunk 4]
```diff
@@ -179,7 +179,7 @@ public void Ack(AckMessage ackMessage)
     {
         // TODO: what if ackMessage.SequenceId is larger than last sent message?
```
Comment (mitchdenny) [inline src/SignalR/common/Shared/MessageBuffer.cs:180]: Does this need to be addressed?

File: src/SignalR/server/Core/src/HubConnectionContextOptions.cs (hunk 3 above)
Comment (mitchdenny) [reply src/SignalR/server/Core/src/HubConnectionContextOptions.cs:41]: So does that mean you pushed back ;)

File: src/SignalR/server/SignalR/test/Internal/MessageBufferTests.cs [hunk 5]
```diff
@@ -235,6 +235,43 @@ public async Task WriteManyMessagesAckSomeProperlyBuffers()
             pipes.Application.Input.AdvanceTo(buffer.Start);
         }
     }
+
+    [Fact]
+    public async Task MessageBufferLimitCanBeModified()
```
Comment (mitchdenny) [inline src/SignalR/server/SignalR/test/Internal/MessageBufferTests.cs:240]: I might be missing it but I'm not seeing where the message buffer limit is modified here.

**Review by mitchdenny — COMMENTED** [review]
A few questions.

File: src/SignalR/server/SignalR/test/Internal/MessageBufferTests.cs (hunk 5 above)
Comment (mitchdenny) [reply src/SignalR/server/SignalR/test/Internal/MessageBufferTests.cs:240]: Oh wait ... you are changing it from the default :) Duh!.

**Review by mitchdenny — APPROVED** [review]

File: src/SignalR/common/Shared/MessageBuffer.cs (hunk 4 above)
Comment (BrennanConroy) [reply src/SignalR/common/Shared/MessageBuffer.cs:180]: I don't think it's important. It's more like provide an error for custom clients that implement acking incorrectly.

**Pull request merged**